File: week4/p3.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* prev;
    struct Node* next;
};

struct Node* reverseDLL(struct Node* head) {
    struct Node* temp = NULL;
    struct Node* curr = head;
    while (curr) {
        temp = curr->prev;
        curr->prev = curr->next;
        curr->next = temp;
        curr = curr->prev;
    }
    if (temp) head = temp->prev;
    return head;
}

void printList(struct Node* head) {
    while (head) {
        printf("%d", head->data);
        if (head->next) printf(" <-> ");
        head = head->next;
    }
    printf("\n");
}

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

int main() {
    struct Node* head = createNode(3);
    struct Node* n2 = createNode(4);
    struct Node* n3 = createNode(5);

    head->next = n2;
    n2->prev = head;
    n2->next = n3;
    n3->prev = n2;

    printf("Original list:\n");
    printList(head);

    head = reverseDLL(head);

    printf("Reversed list:\n");
    printList(head);

    return 0;
}

File: week4/p3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

def reverseDLL(head):
    if not head:
        return None
    curr = head
    temp = None
    while curr:
        temp = curr.prev
        curr.prev = curr.next
        curr.next = temp
        curr = curr.prev
    if temp:
        head = temp.prev
    return head

def printList(head):
    while head:
        print(head.data, end=" <-> " if head.next else "")
        head = head.next
    print()

if __name__ == "__main__":
    head = Node(3)
    n2 = Node(4)
    n3 = Node(5)
    head.next = n2
    n2.prev = head
    n2.next = n3
    n3.prev = n2

    print("Original list:")
    printList(head)

    head = reverseDLL(head)

    print("Reversed list:")
    printList(head)
